pmlc/dialect/pxa/transforms/stencil.cc

```diff
@@ -253,8 +304,204 @@ void Stencil::CollectStrideOneIndices() {
+           IsStrideOne(idx, tensorsOrder[2]);
+  case 1:
+    return true;
+  case 2:
+    return IsStrideOne(idx, tensorsOrder[0]);
+  default:
+    llvm_unreachable("Wrong matrix_idx");
+  }
+  return false;
+}
+
+bool Stencil::IndexExists(mlir::BlockArgument idx, unsigned tensor_idx) {
+  return tensorIdxs[tensor_idx].find(idx) != tensorIdxs[tensor_idx].end();
+}
+
+// Confirm if idx exists in the right place
+bool Stencil::ValidateIndexExistance(mlir::BlockArgument idx,
+                                     unsigned matrix_idx) {
+  switch (matrix_idx) {
+  case 0:
+    // Test if M exists in B and C, does not exist in A
+    return !IndexExists(idx, tensorsOrder[0]) && //
+           IndexExists(idx, tensorsOrder[1]) &&  //
+           IndexExists(idx, tensorsOrder[2]);
+  case 1:
+    // Test if N exists in A and C, does not exist in B
+    return IndexExists(idx, tensorsOrder[0]) &&  //
+           !IndexExists(idx, tensorsOrder[1]) && //
+           IndexExists(idx, tensorsOrder[2]);
+  case 2:
+    // Test if K exists in A and B, does not exist in C
+    return IndexExists(idx, tensorsOrder[0]) && //
+           IndexExists(idx, tensorsOrder[1]) && //
+           !IndexExists(idx, tensorsOrder[2]);
+  default:
+    llvm_unreachable("Wrong matrix_idx.");
+  }
+  return false;
+}
+
+// Search for matrix index (0 for M, 1 for N, 2 for K)
+void Stencil::SearchIndex(unsigned matrix_idx) {
+  if (matrix_idx >= kNumIndex) {
+    // We have the index and then search the tiles for these index
+    SearchTiles(0);
+    return;
+  }
+  auto &idxs = (matrix_idx == kNumIndex - 1) ? allIdxs : outIdxs;
+  for (auto idx : idxs) {
+    if (ValidateStrideOne(idx, matrix_idx) &&
+        ValidateIndexExistance(idx, matrix_idx)) {
+      innerIdxs[matrix_idx] = idx;
+      SearchIndex(nextMatrixIdx[matrix_idx]);
+    }
+  }
+}
+
+void Stencil::SearchTensorsOrder() {
+  // A B C, Search M(0) first as M is most restricted index
+  tensorsOrder[0] = 0;
+  tensorsOrder[1] = 1;
+  tensorsOrder[2] = 2;
+  SearchIndex(0);
+  // B A C, Search M(0) first as M is most restricted index
+  tensorsOrder[0] = 1;
+  tensorsOrder[1] = 0;
+  SearchIndex(0);
+}
+
+void Stencil::SearchTiles(unsigned idx) {
+  if (idx >= kNumIndex) {
+    double performance = Evaluate();
+    if (performance < bestPerf) {
+      bestPerf = performance;
+      for (unsigned i = 0; i < kNumTensors; ++i) {
+        bestTensorsOrder[i] = tensorsOrder[i];
+      }
+      for (unsigned i = 0; i < kNumIndex; ++i) {
+        bestIdxs[i] = innerIdxs[i];
+        bestTiles[i] = tiles[i];
+      }
+    }
+    return;
+  }
+
+  unsigned range = idxRange(innerIdxs[idx]);
+  for (unsigned i = range; i > 0; --i) {
+    if (onlyEven[idx] && (range % i != 0)) {
+      continue;
+    }
+    tiles[idx] = i;
+    SearchTiles(idx + 1);
+  }
+}
+
+int64_t Stencil::idxRange(mlir::BlockArgument idx) {
+  if (opConstRanges != llvm::None) {
+    return (*opConstRanges)[idx.getArgNumber()];
+  }
+  return -1;
```

Comment:
This should be an assertible condition, returning -1 here makes no sense.  Also, we probably shouldn't even use an optional as the member variable, since if the getConstantRanges() ranges fail, we exit early.  We should probably make opConstRanges just a SmallVec, and if getConstantRanges() succeeds, assign it by value.  Might also be good to assert that idx. getArgNumber() < opConstRanges.size().

Reply:
Not using the Optional anymore.